diff --git a/templates/nix/crate2nix/default.nix b/templates/nix/crate2nix/default.nix
index 4eefda8..d064118 100644
--- a/templates/nix/crate2nix/default.nix
+++ b/templates/nix/crate2nix/default.nix
@@ -111,6 +111,7 @@ rec {
             (
               _: {
                 buildTests = true;
+                release = false;
               }
             );
           # If the user hasn't set any pre/post commands, we don't want to
